storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java

```diff
@@ -1984,11 +2074,13 @@ private int fragmentedCpu() {
         Cluster cluster = new Cluster(inimbus, supervisors, topoToSchedAssignment, topologies, conf);
         cluster.setStatusMap(idToSchedStatus.get());
 
-        long beforeSchedule = System.currentTimeMillis();
+        schedulingStartTime.set(Time.nanoTime());
         scheduler.schedule(topologies, cluster);
-        long scheduleTimeElapsedMs = System.currentTimeMillis() - beforeSchedule;
-        LOG.debug("Scheduling took {} ms for {} topologies", scheduleTimeElapsedMs, topologies.getTopologies().size());
-        scheduleTopologyTimeMs.update(scheduleTimeElapsedMs);
+        //Will compiler optimize the order of evalutation and cause race condition?
```

Comment:
This is probably trivial, but I think it's possible to have longestSchdulingTime slightly higher max of scheduleTopologyTimeMs.

Reply:
How?

Reply:
If compiler is allowed to hoist code, we can ended up with the case: 
1) gauge "longest-scheduling-time-ms" acquires scheduling start time 
2) elpased capture end time, update `schedulingDuration` Timer and `longestSchedulingTime`
3) gauge acquires longest time from `longestSchedulingTime` and `currTime` 
4) since start time isn't null, program execute statement 

```java
longest = currTime - startTime > longest ? currTime - startTime : longest;
```
So it'll be slightly longer than the value of `longestSchedulingTime` for this round. Which means we might saw jiggles in gauge value

Reply:
Right, that makes sense. I don't think it requires reordering though, the sequence you posted can happen without the compiler reordering anything.

I don't think it really hurts anything to have this race, but maybe we should consider removing the longest update in the gauge. Unless the scheduler is outright hanging, I'm not sure what the value is for us to report a partial measurement like that?

Reply:
If we keep the update and race, I think we need to either remove the comments about the race, or expand them to provide the example you posted. Leaving the comment as is won't help someone reading it later understand what race is there.

Reply:
If no code reordering happens, gauge should always evaluate `currTime` first, and has to get startTime from `schedulingStartTimeNs` is set to null. So if we guarantee that elpased is evaluated after that, longest-scheduling-time-ms will not exceeds the real longest scheduling time.

That being said, I think the race here should be pretty negligible especially if we discard the decimals in ns-to-ms conversion. Meanwhile I did hear about complaints of hanging schedulers before, so I say we keep partial measurement and remove the comments, or just "please be noticed that it's normal to see minor jiggling in the longest scheduling time due to race condition."

Reply:
You're right. I don't see a good fix. I agree that the race is not that important, I'd maybe just remove the comments, since users won't see them here.